### PR TITLE
[GH-583] Add daily rewards

### LINKED ISSUE #583: [FEATURE] Daily rewards (login streak)
https://www.notion.so/lambdaclass/Daily-Login-9b9db0c820bd49f29b5c75753ad7205b

### PR DESCRIPTION
**Pull request by Nico-Sanchez**
## Motivation

We need a daily rewards feature.
Closes #583
More info at -> https://www.notion.so/lambdaclass/Daily-Login-9b9db0c820bd49f29b5c75753ad7205b

## Summary of changes

- [Add functions for daily rewards logic](https://github.com/lambdaclass/mirra_backend/pull/602/commits/0095ebed5a2dcafd5f4d8b0ba25cf2431c7b96ce)
- [Add daily rewards business rules json and config](https://github.com/lambdaclass/mirra_backend/pull/602/commits/80486f54a00933d6c39ae267bcb4db7d8c3a9fd2)
- [Add daily rewards fields to user table and schema](https://github.com/lambdaclass/mirra_backend/pull/602/commits/4ff2db518337e8b36f2354593b9a6e91d240bd71)
- [Add endpoint to claim current daily reward](https://github.com/lambdaclass/mirra_backend/pull/602/commits/5b32c430b519c3121dcb1899e97cf9490caf841d)
- [Add new endpoint to get the user daily reward status](https://github.com/lambdaclass/mirra_backend/pull/602/commits/045fa27aa689977f04f7b80bce8c801ff508e3fd)

## How to test it?

1. Insert a user `{:ok, user} = GameBackend.Users.find_or_create_google_user_by_email("a")`
2. `user_id = user.user.id`
3. GET request to: `/users/:user_id/claim_daily_reward` and claim the reward for today:
  a. `curl -X GET http://localhost:4001/users/:user_id/claim_daily_reward`
4. Then you can check your rewards status with:
  a. `curl -X GET http://localhost:4001/users/:user_id/get_daily_reward_status`

If you want to keep claiming rewards, you have to be on a streak. To do so you can edit the user's data:
```elixir
Users.update_user(Users.get_user(user_id) |> elem(1), %{
             last_daily_reward_claim_at: DateTime.utc_now() |> DateTime.add(-1, :day),
             last_daily_reward_claim: "day_to_claim"
           })
``` 

## Checklist
- [x] Tested the changes locally.
- [x] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: apps/game_backend/lib/game_backend/rewards.ex [hunk 1]
```diff
@@ -43,4 +43,45 @@ defmodule GameBackend.Rewards do
...
+  @doc """
+  Receives a user.
+  Returns {:ok, can_claim} if the user claimed today already.
+  """
+  def user_claimed_today?(user) do
+    now = DateTime.utc_now()
+
+    case Date.compare(user.last_daily_reward_claim, now) do
+      :eq -> {:ok, :can_claim}
+      _ -> {:error, :already_claimed}
```
Comment (AminArria) [inline apps/game_backend/lib/game_backend/rewards.ex:84]: This return values are unexpected for Elixir conventions of functions ending with `?`, any function like that should return a boolean

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 2]
```diff
@@ -14,4 +15,22 @@ defmodule Gateway.Controllers.UserController do
...
+             last_daily_reward_claim: DateTime.utc_now(),
+             last_daily_reward_claim_type: daily_reward
+           }) do
+      send_resp(conn, 200, Jason.encode!(user.id))
+    else
+      {:error, :not_found} -> send_resp(conn, 404, Jason.encode!(%{"error" => "not found"}))
+      {:error, :already_claimed} -> send_resp(conn, 400, Jason.encode!(%{"error" => "already claimed"}))
+      {:error, :invalid_reward} -> send_resp(conn, 400, Jason.encode!(%{"error" => "invalid reward"}))
+      {:error, _changeset} -> send_resp(conn, 400, Jason.encode!(%{"error" => "failed to update"}))
+    end
```
Comment (AminArria) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:34]: Let's try and improve this

First of ["Anti-pattern: complex else in with"](https://hexdocs.pm/elixir/code-anti-patterns.html#complex-else-clauses-in-with) for a brief explanation on why doing `with` like this is not a good idea for maintainability

Second ["Phoenix: Action fallback"](https://hexdocs.pm/phoenix/json_and_apis.html#action-fallback) explaining a more generic approach to dealing with this kind of things. Also good to read the following section ["FallbackController and ChangesetJSON"](https://hexdocs.pm/phoenix/json_and_apis.html#fallbackcontroller-and-changesetjson)

This means that ideally this function should end up more like
```elixir
  def claim_daily_reward(conn, %{"user_id" => user_id}) do
    with {:ok, user} <- Users.get_user(user_id),
         {:ok, :can_claim} <- Rewards.user_claimed_today?(user),
         {:ok, daily_reward} <- Rewards.claim_daily_reward(user),
         {:ok, user} <- Users.update_user(user, %{last_daily_reward_claim: DateTime.utc_now(), last_daily_reward_claim_type: daily_reward}) do
      send_resp(conn, 200, Jason.encode!(user.id))
    end
  end
```

File: apps/game_backend/priv/repo/migrations/20240510172559_add_daily_rewards_fields_to_user.exs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule GameBackend.Repo.Migrations.AddDailyRewardsFieldsToUser do
+  use Ecto.Migration
+
+  def change do
+    alter(table(:users)) do
+      add(:last_daily_reward_claim, :utc_datetime)
```
Comment (AminArria) [inline apps/game_backend/priv/repo/migrations/20240510172559_add_daily_rewards_fields_to_user.exs:6]: Please put `_at` at the end

**Review by AminArria — CHANGES_REQUESTED** [review]

File: apps/game_backend/lib/game_backend/users/user.ex [hunk 4]
```diff
@@ -38,6 +40,8 @@ defmodule GameBackend.Users.User do
       :game_id,
       :username,
       :last_afk_reward_claim,
+      :last_daily_reward_claim,
+      :last_daily_reward_claim_type,
```
Comment (AminArria) [inline apps/game_backend/lib/game_backend/users/user.ex:44]: What does this mean?

File: apps/game_backend/priv/daily_rewards_rules.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+    "daily_rewards_system": {
+        "1": "2",
+        "2": "3",
+        "3": "4",
+        "4": "5",
+        "5": "6",
+        "6": "7",
+        "7": "1"
```
Comment (AminArria) [inline apps/game_backend/priv/daily_rewards_rules.json:9]: What happens after the 7th day?

File: apps/game_backend/lib/game_backend/rewards.ex (hunk 1 above)
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/rewards.ex:84]: Yeah thanks Amin.

File: apps/game_backend/priv/repo/migrations/20240510172559_add_daily_rewards_fields_to_user.exs (hunk 3 above)
Comment (Nico-Sanchez) [reply apps/game_backend/priv/repo/migrations/20240510172559_add_daily_rewards_fields_to_user.exs:6]: I did that first, but changed to match the other reward field. Since we agree is better to do this, I'll change it.

File: apps/game_backend/lib/game_backend/users/user.ex (hunk 4 above)
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/users/user.ex:44]: It's to know when a user claimed their last daily reward.

File: apps/game_backend/priv/daily_rewards_rules.json (hunk 5 above)
Comment (Nico-Sanchez) [reply apps/game_backend/priv/daily_rewards_rules.json:9]: Goes back to first one 😄 
https://www.notion.so/lambdaclass/Daily-Login-9b9db0c820bd49f29b5c75753ad7205b

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 2 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex]: You're totally right about the with-else and I didn't know about the other two articles. I'll implement the 1st then read the others. Thanks a lot!

**Comment by manucamejo** [comment]
Good job @Nico-Sanchez.
Can you create an endpoint that provides information about the quests and where you are?
Like a list of: [{day, rewards, status}, ...]

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 6]
```diff
@@ -4,14 +4,27 @@ defmodule Gateway.Controllers.UserController do
...
     with {:ok, user} <- Users.get_user(params["user_id"]),
          {:ok, user} <- Users.update_user(user, params) do
       send_resp(conn, 200, Jason.encode!(user.id))
-    else
-      {:error, :not_found} -> send_resp(conn, 404, Jason.encode!(%{"error" => "not found"}))
-      {:error, _changeset} -> send_resp(conn, 400, Jason.encode!(%{"error" => "failed to update"}))
+    end
+  end
+
+  def claim_daily_reward(conn, %{"user_id" => user_id}) do
```
[the remark this answers is not in the document]
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex:19]: Yeah. It's for the users to claim today's reward.

**Comment by Nico-Sanchez** [comment]
> Good job @Nico-Sanchez. Can you create an endpoint that provides information about the quests and where you are? Like a list of: [{day, rewards, status}, ...]

Alright. Done!

File: apps/game_backend/lib/game_backend/rewards.ex [hunk 7]
```diff
@@ -43,4 +44,73 @@ defmodule GameBackend.Rewards do
...
+
+  defp user_in_daily_rewards_streak?(user) do
+    today = DateTime.utc_now()
+    yesterday = today |> Date.add(-1)
+
+    Date.compare(user.last_daily_reward_claim_at, today) == :eq or
+      Date.compare(user.last_daily_reward_claim_at, yesterday) == :eq
+  end
+
+  def mark_user_daily_rewards_as_completed(daily_rewards_config, user) do
```
Comment (AminArria) [inline apps/game_backend/lib/game_backend/rewards.ex:105]: Not sure what is wrong, but something is wrong here cause I get the following (or I missunderstood something)

```
>curl http://localhost:4001/users/bd984c53-ca09-418e-aeb7-9ba6cbf58f54/get_daily_reward_status | jq
{
  "1": "not claimed",
  "2": "not claimed",
  "3": "not claimed",
  "4": "not claimed",
  "5": "not claimed",
  "6": "not claimed",
  "7": "not claimed"
}

> curl http://localhost:4001/users/bd984c53-ca09-418e-aeb7-9ba6cbf58f54/claim_daily_reward

> curl http://localhost:4001/users/bd984c53-ca09-418e-aeb7-9ba6cbf58f54/get_daily_reward_status | jq
{
  "1": "not claimed", ## <- this should be claimed right?
  "2": "claimed",       ## <- This and all other should be not claimed right?
  "3": "claimed",
  "4": "claimed",
  "5": "claimed",
  "6": "claimed",
  "7": "claimed"
}
```

File: apps/game_backend/priv/daily_rewards_rules.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
+    "reward_per_day": {
+        "1": "2",
```
Comment (AminArria) [inline apps/game_backend/priv/daily_rewards_rules.json:3]: Is this supposed to be `day: next_day` or `day: reward_for_day`?

**Review by AminArria — CHANGES_REQUESTED** [review]
Can you elaborate on how to test the full cycle of rewards? I want to test giving my self all daily rewards, but modifying `last_daily_reward_claim_at` didn't seem to do the trick

File: apps/game_backend/lib/game_backend/rewards.ex [hunk 9]
```diff
@@ -43,4 +44,73 @@ defmodule GameBackend.Rewards do
...
+  Receives a user.
+  Returns {:ok, can_claim} if the user claimed today already or never claimed at all.
+  """
+  def user_claimed_today(%GameBackend.Users.User{last_daily_reward_claim_at: nil}), do: {:ok, :can_claim}
+
+  def user_claimed_today(user) do
+    now = DateTime.utc_now()
+
+    case Date.compare(user.last_daily_reward_claim_at, now) do
+      :eq -> {:ok, :can_claim}
```
Comment (AminArria) [inline apps/game_backend/lib/game_backend/rewards.ex:90]: Is this correct? Shouldn't this be `:lt` so it says how you haven't claimed today

File: apps/game_backend/lib/game_backend/rewards.ex (hunk 7 above)
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/rewards.ex:108]: This should be fixed now!

File: apps/game_backend/priv/daily_rewards_rules.json (hunk 8 above)
Comment (Nico-Sanchez) [reply apps/game_backend/priv/daily_rewards_rules.json]: Since its hard to follow, I reworked this config to be more descriptive.

File: apps/game_backend/lib/game_backend/rewards.ex (hunk 9 above)
Comment (Nico-Sanchez) [reply apps/game_backend/lib/game_backend/rewards.ex]: Yeah you're right, fixed now.

**Review by AminArria — DISMISSED** [review]

File: config/runtime.exs [hunk 10]
```diff
@@ -146,6 +146,12 @@ if System.get_env("RELEASE") == "central_backend" or config_env() == :dev do
   config :game_backend, :currencies_config, Jason.decode!(currency_config_json)
 end
 
+{:ok, daily_rewards_json} =
```
Comment (AminArria) [inline config/runtime.exs:149]: Please try running a `arena` release as this might break it

Comment (Nico-Sanchez) [reply config/runtime.exs]: Done. Nice catch!

**Review by AminArria — DISMISSED** [review]

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 11]
```diff
@@ -4,14 +4,47 @@ defmodule Gateway.Controllers.UserController do
...
+             last_daily_reward_claim_at: DateTime.utc_now(),
+             last_daily_reward_claim: daily_reward["day"]
+           }),
+         {:ok, %UserCurrency{}} <-
+           Currencies.add_currency_by_name_and_game!(
+             user_id,
+             daily_reward["currency"],
+             Utils.get_game_id(:curse_of_mirra),
+             daily_reward["amount"]
+           ) do
```
Comment (agustinesco) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:36]: Let's do this operation in a multi, i think this is error-prone and users may miss daily rewards if the operations fails

File: apps/gateway/lib/gateway/controllers/user_controller.ex [hunk 12]
```diff
@@ -4,14 +4,47 @@ defmodule Gateway.Controllers.UserController do
...
+      send_resp(conn, 200, Jason.encode!(user.id))
+    end
+  end
+
+  def get_daily_reward_status(conn, %{"user_id" => user_id}) do
+    with {:ok, user} <- Users.get_user(user_id) do
+      user_daily_reward_status =
+        Utils.get_daily_rewards_config()
+        |> Rewards.mark_user_daily_rewards_as_completed(user)
```
Comment (agustinesco) [inline apps/gateway/lib/gateway/controllers/user_controller.ex:45]: Why are we marking rewards as completed when we check the status?

File: apps/gateway/lib/gateway/router.ex [hunk 13]
```diff
@@ -27,6 +27,8 @@ defmodule Gateway.Router do
     get "/auth/:provider/token/:token_id", Controllers.AuthController, :validate_token
 
     put "/users/:user_id", Controllers.UserController, :update
+    get "/users/:user_id/claim_daily_reward", Controllers.UserController, :claim_daily_reward
+    get "/users/:user_id/get_daily_reward_status", Controllers.UserController, :get_daily_reward_status
```
Comment (agustinesco) [inline apps/gateway/lib/gateway/router.ex:31]: I think these routes should be under the "curse" scope since these are operations that belongs there

**Review by agustinesco — CHANGES_REQUESTED** [review]
The reward feature work as expected, nice! let's solve some thing about the code and we're good to go

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 12 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex:45]: It's a part of building the status. I don't store it, I calculate it everytime.

File: apps/gateway/lib/gateway/router.ex (hunk 13 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/router.ex:31]: Alright. Done.

File: apps/gateway/lib/gateway/controllers/user_controller.ex (hunk 11 above)
Comment (Nico-Sanchez) [reply apps/gateway/lib/gateway/controllers/user_controller.ex]: Alright, done!

**Review by agustinesco — APPROVED** [review]
LGTM!
